extension/httpfs/hffs.cpp

```diff
@@ -0,0 +1,407 @@
+			result.push_back(GetHFUrl(curr_hf_path));
+		}
+	}
+
+	// Prune files using match
+	return result;
+}
+
+unique_ptr<ResponseWrapper> HuggingFaceFileSystem::HeadRequest(FileHandle &handle, string hf_url,
+                                                               HeaderMap header_map) {
+	auto &hf_handle = handle.Cast<HFFileHandle>();
+	auto http_url = HuggingFaceFileSystem::GetFileUrl(hf_handle.parsed_url);
+	return HTTPFileSystem::HeadRequest(handle, http_url, header_map);
+}
+
+unique_ptr<ResponseWrapper> HuggingFaceFileSystem::GetRequest(FileHandle &handle, string s3_url, HeaderMap header_map) {
+	auto &hf_handle = handle.Cast<HFFileHandle>();
+	auto http_url = HuggingFaceFileSystem::GetFileUrl(hf_handle.parsed_url);
+	return HTTPFileSystem::GetRequest(handle, http_url, header_map);
+}
+
+unique_ptr<ResponseWrapper> HuggingFaceFileSystem::GetRangeRequest(FileHandle &handle, string s3_url,
+                                                                   HeaderMap header_map, idx_t file_offset,
+                                                                   char *buffer_out, idx_t buffer_out_len) {
+	auto &hf_handle = handle.Cast<HFFileHandle>();
+	auto http_url = HuggingFaceFileSystem::GetFileUrl(hf_handle.parsed_url);
+	return HTTPFileSystem::GetRangeRequest(handle, http_url, header_map, file_offset, buffer_out, buffer_out_len);
+}
+
+unique_ptr<HTTPFileHandle> HuggingFaceFileSystem::CreateHandle(const string &path, FileOpenFlags flags,
+                                                               optional_ptr<FileOpener> opener) {
+	D_ASSERT(flags.Compression() == FileCompressionType::UNCOMPRESSED);
+
+	auto parsed_url = HFUrlParse(path);
+
+	auto params = HTTPParams::ReadFrom(opener);
+	SetParams(params, path, opener);
+
+	return duckdb::make_uniq<HFFileHandle>(*this, std::move(parsed_url), path, flags, params);
+}
+
+void HuggingFaceFileSystem::SetParams(HTTPParams &params, const string &path, optional_ptr<FileOpener> opener) {
+	auto secret_manager = FileOpener::TryGetSecretManager(opener);
+	auto transaction = FileOpener::TryGetCatalogTransaction(opener);
+	if (secret_manager && transaction) {
+		auto secret_match = secret_manager->LookupSecret(*transaction, path, "huggingface");
+
+		if (secret_match.HasMatch()) {
+			const auto &kv_secret = dynamic_cast<const KeyValueSecret &>(*secret_match.secret_entry->secret);
+			params.bearer_token = kv_secret.TryGetValue("token", true).ToString();
+		}
+	}
+}
+
+ParsedHFUrl HuggingFaceFileSystem::HFUrlParse(const string &url) {
+	ParsedHFUrl result;
+
+	if (!StringUtil::StartsWith(url, "hf://")) {
+		throw InternalException("Not an hf url");
+	}
+
+	size_t last_delim = 5;
+	size_t curr_delim;
+
+	// Parse Repository type
+	curr_delim = url.find('/', last_delim);
+	if (curr_delim == string::npos) {
+		throw IOException("URL needs to contain a '/' after the repository type: (%s)", url);
+	}
+	result.repo_type = url.substr(last_delim, curr_delim - last_delim);
```

Comment:
A subtlety here is that, for historical reasons, the [models](https://huggingface.co/models) don't have the `models/` prefix. See https://huggingface.co/HuggingFaceM4/idefics2-8b for example.

I'm not sure if it's crucial, as we want to support hf:// for [datasets](https://huggingface.co/datasets).

Reply:
@severo I've added code that throws an exception if the repository type is NOT `datasets`. Ensuring the user gets a clean error message if anything else is attempted. Does that work? Or should other urls than `hf://datasets/*` be supported too?

Reply:
You can allow `hf://spaces/*` as well

Reply:
ah makes sense, i've added it along with a test. Thanks!